model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/idp/InfinispanIdentityProviderStorageProvider.java

```diff
@@ -140,6 +151,40 @@ public IdentityProviderModel getByAlias(String alias) {
         return createOrganizationAwareIdentityProviderModel(cached.getIdentityProvider());
     }
 
+    @Override
+    public Stream<IdentityProviderModel> getByOrganization(String orgId, Integer first, Integer max) {
+        String cacheKey = cacheKeyOrgId(getRealm(), orgId);
+
+        if (isInvalid(cacheKey)) {
+            return idpDelegate.getByOrganization(orgId, first, max).map(this::createOrganizationAwareIdentityProviderModel);
+        }
+
+        RealmCacheManager cache = realmCache.getCache();
+        IdentityProviderListQuery query = cache.get(cacheKey, IdentityProviderListQuery.class);
+        String searchKey = Optional.ofNullable(first).orElse(-1) + "." + Optional.ofNullable(max).orElse(-1);
+        Set<String> cached;
+
+        if (query == null) {
+            // not cached yet
+            Long loaded = cache.getCurrentRevision(cacheKey);
+            cached = idpDelegate.getByOrganization(orgId, first, max).map(IdentityProviderModel::getInternalId).collect(Collectors.toSet());
+            query = new IdentityProviderListQuery(loaded, cacheKey, getRealm(), searchKey, cached);
+            cache.addRevisioned(query, startupRevision);
+        } else {
+            cached = query.getIDPs(searchKey);
+            if (cached == null) {
+                // there is a cache entry, but the current search is not yet cached
+                cache.invalidateObject(cacheKey);
+                Long loaded = cache.getCurrentRevision(cacheKey);
+                cached = idpDelegate.getByOrganization(orgId, first, max).map(IdentityProviderModel::getInternalId).collect(Collectors.toSet());
+                query = new IdentityProviderListQuery(loaded, cacheKey, getRealm(), searchKey, cached, query);
+                cache.addRevisioned(query, cache.getCurrentCounter());
+            }
+        }
+
+        return cached.stream().map(this::getById);
```

Comment:
It might happen an organization is being removed/invalidated so we need to make sure the ids here are valid. Similar to what we have in groups. See org.keycloak.models.cache.infinispan.RealmCacheSession#getGroupsStream(org.keycloak.models.RealmModel).

Reply:
Fixed